### ISSUE TITLE
Shared lockfile results in nondeterministic top-level packages, differing behavior from npm install

### ISSUE THREAD

**Issue by jakebailey** [pnpm/pnpm#6457]
### Code to reproduce the issue:

https://github.com/jakebailey/pnpm-dt-repro-1

### Expected behavior:

This is a little tricky to describe; right now, DT effectively loops over each package and runs `npm install`. This means that every package has its own `node_modules`. If a workspace packages's dependency has an undeclared dependency on another external package, that issue can be resolved by the workspace package. What we're looking for is some sort of middle-ground that can fix the issue below.

### Actual behavior:

All packages get linked into a top-level `.pnpm/node_modules`. But, the contents of that folder are non-deterministic. If multiple `@types/node` packages are present in the package graph, the package that gets linked for "undocumented" dependencies is effectively random based on what other packages happen to do. In my case, `.pnpm/node_modules/@types/node` turns out to be `@types/node@v17`, which breaks things. @sandersn has exprienced it being a totally different version. In my repro, you can remove `gsheetdb` from the repo and it'll resolve to yet another version.

By switching out of "shared lockfile mode", we avoid the problem, because each workspace package gets its own `node_modules` directory and hoisted `.pnpm/node_modules`, mirroring what you get when you just `npm install`. It also matches yarn's `nmHoistingLimits` option set to "workspace", I think. But, it's really, really slow, as each workspace package seems to be individually recalculating things that could be shared.

I'm not really sure what the right answer is; I don't think "normal" users hit this, but DT is really stretching the limits by using multiple versions of the same package in the same monorepo, even _declaring_ the same package more than once.

### Additional information:

 - `node -v` prints: 19.9.0
 - Windows, macOS, or Linux?: Linux

**Comment by zkochan** [comment]
So why don't you just turn off the hoisting to `node_modules/.pnpm/node_modules`? You can set `hoist=false`.

**Comment by jakebailey** [comment]
I have tried that, but that then means that any and all undeclared dependencies break without a way to resolve them.

For example, the case I have here is that there's a package (`form-data`) which has written `/// <references types="node" />`. So, this package "depends" on `@types/node`, but doesn't declare that, which is really common for typed packages (as they are typically always dev deps).

That package is placed within `/node_modules/.pnpm/form-data+3.0.1`. So, when it goes and looks up the import for `node`, it walks up the dirs and never finds `node`, so fails. In the case where `hoist=true`, the same process happens, except that it finds whatever random version of `@types/node` happened to be hoisted.

If that same code had instead been linked into the workspace package that used it, it would have instead been like:

```
types/node-fetch/node_modules/.pnpm/form-data+3.0.1
types/node-fetch/node_modules/@types/node -> types/node
```

And then walking up finds a version of `@types/node`, which is what _would_ have happened if `@types/node-fetch` had been installed in a user's codebase.

**Comment by jakebailey** [comment]
I'm having trouble coming up with a "good" solution, unfortunately.

Another way to "fix this" is to configure TS to not resolve symlinks. This gives us resolution that mostly matches what a plain install would look like because then resolution comes from the correct `node_modules`, but, then self-linking breaks because that's _also_ a symlink!

It could be that a mode that hardlinks dependencies rather than indirecting through `.pnpm` would work, but then now, it's technically the case that dependencies are duplicated in multiple folders and are loaded by TS. But, that's what we already have now with the naive "loop over and `npm install`" method we're trying to move away from.

I'm going to try and see if there are any quick performance wins for a non-shared lockfile system, at least, given that gives us the closest result to the `npm` layout.

I'll also mention that I believe that `yarn` hoisting limits also give us a pretty good layout; let me make a branch of my repro to double check. There are many things that _don't_ work, but, I think that's a case which shows that a single lockfile is possible while still linking things differently.

**Comment by zkochan** [comment]
The only workaround for this would be to make all direct dependencies peer dependencies of all direct dependencies. But that might make installation slower. And in your case it is a workspace with thousands of projects, so it will be probably terribly slow. It will probably even go out of memory.

**Comment by zkochan** [comment]
However, if you could find which packages are actually referenced using code analysis and pass it to a pnpm hook that would make those packages peer dependencies. That would be probably fast. Except the code analysis part I guess.

**Comment by jakebailey** [comment]
Is there any way to provide control over which dependencies get stuck into `.pnpm/node_modules`? e.g., in the current repo, there are versions `[12, 14, 16, 17, 18]`, but which one goes there seems to be arbitrary, but I think should probably be `18`. That might fix the bulk of our problems; 255 of the 562 failures on the DT monorepo WIP branch are just from `@types/node` getting mis-resolved.

**Comment by jakebailey** [comment]
Theoretically, we could just try and do peer dep patches to try and fix some of these... But given it's not an "error" for package to implicitly depend on `@types/node`, it's probably not going to be managable.

**Comment by zkochan** [comment]
Well, if most of the issues are with `@types/node` then you can try to add `@types/node` to peer dependencies of every `@types` package.

**Comment by jakebailey** [comment]
Hm, I suppose that somewhat matches the current non-monorepo which effectively peer deps _all_ `@types` packages (but then sets dependencies based on static analysis pre-publish). We already set `"types": []` to stop them from being loaded. I'll have to try and see if that helps. But, `@types/node` may not be the only offenders.

For sake of comparison, here's the same repro with yarn3, which works: https://github.com/jakebailey/pnpm-dt-repro-1/tree/yarn

There's some problem with self-linking that required a small test change, but it does provide a layout that functions as expected.

**Comment by zkochan** [comment]
pnpm can also implement this:
`nmHoistingLimits: workspaces`

we use Yarn's package for hoisting, so it should be possible to change the hoisting limits. It will be probably much slower though. Probably the similar speed to non-shared lockfile.

**Comment by jakebailey** [comment]
> we use Yarn's package for hoisting, so it should be possible to change the hoisting limits. It will be probably much slower though.

Interesting! I acutally have a WIP branch that has all of DT on yarn3, and it performs similarly to pnpm with this setting set (~1m30s), so, maybe it does work.

I'm totally happy to try it out, if the change is straightforward (or, try and make the change).

**Comment by zkochan** [comment]
If you want to make the change. You can see that here is a hoistingLimits map:

https://github.com/pnpm/pnpm/blob/main/pkg-manager/real-hoist/src/index.ts

I think when that setting is set to workspaces, all the workspace projects should be added to that list. But I don't know in what format.

**Comment by jakebailey** [comment]
Hm, hit an immediate snag; I forgot that `node-linker=hoisted` doesn't work with `lockfile=false`, erroring out with:

```
 ERR_PNPM_OUTDATED_LOCKFILE  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up to date with types/node/scripts/generate-docs/package.json
```

In fact, it also fails when `lockfile=true` on the first run, then suceeds the second time.

This is reproducible with my smaller example, thankfully. Maybe worth a new issue?

**Comment by zkochan** [comment]
I guess it can be fixed.

**Comment by jakebailey** [comment]
Well, for that bug, it comes down to `getSpecFromPackageManifest` returning the empty string and that being added into the snapshot, later making `satisfiesPackageManifest` fail. But, it's eventually dropped and so the resulting lockfile when reloaded doesn't ahve the empty string and thing work.

I'm not totally sure of the root cause of why it's happening; in this case, it's a "script" package which has somehow gained a dependency that is not listed in its `package.json` at all.

A workaround is to always ignore the result of `getSpecFromPackageManifest` when it's falsy, but, that seems like papering over the problem. I can try to come up with some smaller repro, if that is useful (and report this elsewhere).

As for the actual issue of trying to make use of the hoisting levels, the API provided by `hoist` is... weird. It seems like there's a lot of calculation that needs to happen to calculate the hoisting limits, more than just setting one string, and I'm not totally sure how to produce that given it's the hoist that produces the graph. But, at least I'm unblocked via my patch and can look further.

**Comment by jakebailey** [comment]
It was tricky but I finally got an equivalent to `nmHoistingLimits` working; I'm trying to figure out where exactly the hoisting limit calculation needs to live but will send a PR when I have it cleaned up.

**Comment by rrosenshain-sc** [comment]
@jakebailey Was curious if you found a workaround for this or still planning to merge your config change in?
I'm running into the exact same issue.

**Comment by jakebailey** [comment]
I haven't had time to work on my PR in a bit, though my situation is so weird that I doubt your problem is related without a repro 😅

I'm hoping to come back to it but the yarn API here is so confusing that it's a real challenge for me to understand if I'm even doing it right.

**Comment by rrosenshain-sc** [comment]
Reading through this, it sounds very similar though mine is not with `@types` but with `react-intl` (which needs to be a singleton)
I opened a discussion about it a few days ago https://github.com/orgs/pnpm/discussions/7070

Tldr, it seems like pnpm arbitrarily decides which package to hoist to `node_modules/.pnpm/node_modules` when there are multiple options, so if a dependency needs to read from node_modules it may pick different version each time.

It's not deterministic which is the scariest thing because it leads to runtime bugs in production you can't predict.

**Comment by rrosenshain-sc** [comment]
I have a repro example: https://github.com/swang9-sc/pnpm-demo/tree/main
Basically running `pnpm i` multiple times generates a different tree randomly

**Comment by zkochan** [comment]
Resolved by #12041 (merged).

pnpm now has a `hoistingLimits` setting — `none` (default), `workspaces`, or `dependencies` — mirroring yarn's `nmHoistingLimits`. Under `nodeLinker: hoisted`:

- `dependencies` hoists only up to each workspace package's direct dependencies, giving the deterministic, npm-style per-package nesting (`packages/A/node_modules/B/node_modules/C`) that avoids the nondeterministic shared top-level described in this issue.
- `workspaces` hoists only as far as each workspace package.

It revives the approach proposed in #6468. The setting also landed in the pacquet (Rust) port for parity.

---
Written by an agent (Claude Code, claude-opus-4-8).

**Comment by jakebailey** [comment]
Thanks! I had a chance to test it, but it doesn't respect `publicHoistPattern`, unfortunately. I'll have to make a repro.
